### ISSUE TITLE
Result of floating point computations does not match between AVX2 and AVX512

### ISSUE THREAD

**Issue by petrsm** [ispc/ispc#2782]
Hi guys !
We have complex floating point ISPC code, which produces slightly different results for
avx2-i32x8 vs avx512skx-x16 targets. Both targets have exactly same command line and
we are using "--math-lib=fast".

Is this behavior expected ? Or output should be exactly the same ? 

Unfortunately, I have not isolated the issue yet - I am just checking what is expected state.

Thanks !

**Comment by dbabokin** [comment]
It's expected, at least with `--math-lib=fast`. Fast implementation of libs do differ on different ISAs. It's very difficult / impossible to avoid not sacrificing performance.

**Comment by emoon** [comment]
Yes, the compiler has much more leeway to reorder operations when doing `fast-math` so that means that lower fractions of floating point values usually won't add up between targets.

**Comment by petrsm** [comment]
We are not using --fast-math, but 'fast' library of math functions --math-lib.

However, even if I use --math-lib=default, I still get different results from AVX512 path.

Also, I replaced rcp(x) by '1 / x' and rsqrt(x) by '1 / sqrt(x)' and issue persists.

Is this still expected under above conditions ?

**Comment by dbabokin** [comment]
Try `--math-lib=system` to see if this has an effect - it should be slow, but may help you isolate the problem. "default" library depends on ISA quite heavily and it's not designed to be bit-reproducible across ISAs.

Removing `rcp()` and `rsqrt()` should make things more stable.

If it's not a library, then it could be slightly different sequence of FMA instructions due to optimization. You can try `--opt=disable-fma` to experiment.

**Comment by petrsm** [comment]
Great tips ! Thanks !

**Comment by petrsm** [comment]
Unfortunately it is not. At the end, by avoiding usage of rsqrt() and rcp() I managed to make results of computations to match between AVX512 a AVX2, except of few specific functions. I ran out of time for investigation, so for those which does not match, I always use AVX2.

**Comment by pbrubaker** [comment]
Given that we've sorted out that `rsqrt` and `rcp` were the culprits here, I'm going to close this issue.
